ui/plugin-system/src/runtime/template-variables.ts

```diff
@@ -12,28 +12,36 @@
 // limitations under the License.
 
 import { createContext, useContext } from 'react';
+import { VariablesState } from '@perses-dev/core';
 
-export interface TemplateVariables {
-  variables: Record<string, VariableState>;
-}
-
-/**
- * The value and options for a template variable.
- */
-export interface VariableState {
-  value: string | string[];
-  options?: string[];
-}
+export type TemplateVariableSrv = {
+  state: VariablesState;
+};
 
-export const TemplateVariablesContext = createContext<TemplateVariables | undefined>(undefined);
+export const TemplateVariableContext = createContext<TemplateVariableSrv | undefined>(undefined);
 
-/**
- * Gets TemplateVariables at runtime.
- */
-export function useTemplateVariables(): TemplateVariables {
-  const ctx = useContext(TemplateVariablesContext);
+function useTemplateVariableContext() {
+  const ctx = useContext(TemplateVariableContext);
   if (ctx === undefined) {
-    throw new Error('No TemplateVariablesContext found. Did you forget a Provider?');
+    throw new Error('No TemplateVariableContext found. Did you forget a Provider?');
   }
   return ctx;
 }
+
+export function useTemplateVariableValues(names?: string[]) {
+  const { state } = useTemplateVariableContext();
+
+  if (names === undefined) {
+    return state;
+  }
+
+  const values: VariablesState = {};
+  names.forEach((name) => {
+    const s = state[name];
+    if (s) {
+      values[name] = s;
+    }
+  });
```

Comment:
Do you think there's any value in memoizing this? (I always wonder when I see a loop like this to produce a value.)

Reply:
Yeah definitely, I'm thinking of adding in all the memoization optimizations in a follow to get the flow of logic correct first.